cmd/util/ledger/util/payload_snapshot.go

```diff
@@ -123,27 +126,139 @@ func (p PayloadSnapshot) ApplyChangesAndGetNewPayloads(
 type MapBasedPayloadSnapshot struct {
 	reverseMap map[flow.RegisterID]int
 	payloads   []*ledger.Payload
+	log        zerolog.Logger
 }
 
 var _ MigrationStorageSnapshot = (*MapBasedPayloadSnapshot)(nil)
 
-func NewMapBasedPayloadSnapshot(payloads []*ledger.Payload) (*MapBasedPayloadSnapshot, error) {
+func NewMapBasedPayloadSnapshot(log zerolog.Logger, payloads []*ledger.Payload) (*MapBasedPayloadSnapshot, error) {
+	return NewMapBasedPayloadSnapshotWithWorkers(log, payloads, 1)
+}
+
+// NewMapBasedPayloadSnapshotWithWorkers creates a new MapBasedPayloadSnapshot with the given number of workers.
+// The workers are used to create the reverse map in parallel.
+func NewMapBasedPayloadSnapshotWithWorkers(log zerolog.Logger, payloads []*ledger.Payload, workers int) (*MapBasedPayloadSnapshot, error) {
+
+	// limit the number of workers to prevent too many small jobs
+	const minPayloadsPerWorker = 10000
+	if len(payloads)/minPayloadsPerWorker < workers {
+		workers = len(payloads) / minPayloadsPerWorker
+	}
+	if workers < 1 {
+		workers = 1
+	}
+
+	// log the creation of the snapshot if it is large
+	if len(payloads) > minPayloadsPerWorker {
+		start := time.Now()
+		log.Info().
+			Int("payloads", len(payloads)).
+			Int("workers", workers).
+			Msgf("Creating payload snapshot")
+
+		defer func() {
+			log.Info().
+				Int("payloads", len(payloads)).
+				Int("workers", workers).
+				Dur("duration", time.Since(start)).
+				Msgf("Created payload snapshot")
+		}()
+	}
+
 	payloadsCopy := make([]*ledger.Payload, len(payloads))
 	copy(payloadsCopy, payloads)
-	l := &MapBasedPayloadSnapshot{
-		reverseMap: make(map[flow.RegisterID]int, len(payloads)),
-		payloads:   payloadsCopy,
+
+	type result struct {
+		minimap map[flow.RegisterID]int
+		err     error
 	}
-	for i, payload := range payloadsCopy {
-		key, err := payload.Key()
-		if err != nil {
-			return nil, err
+
+	minimaps := make(chan result, workers)
+	defer close(minimaps)
+
+	// create the minimaps in parallel
+	for i := 0; i < workers; i++ {
+		start := i * len(payloadsCopy) / workers
+		end := (i + 1) * len(payloadsCopy) / workers
+		if end > len(payloadsCopy) {
+			end = len(payloadsCopy)
 		}
-		id, err := convert.LedgerKeyToRegisterID(key)
-		if err != nil {
-			return nil, err
+		go func(payloads []*ledger.Payload) {
+			minimap := make(map[flow.RegisterID]int, len(payloads))
+			for i, payload := range payloads {
+				key, err := payload.Key()
+				if err != nil {
+					minimaps <- result{nil, err}
+					return
+				}
+				id, err := convert.LedgerKeyToRegisterID(key)
+				if err != nil {
+					minimaps <- result{nil, err}
+					return
+				}
+				minimap[id] = i
+			}
+			minimaps <- result{minimap, nil}
+		}(payloadsCopy[start:end])
```

Comment:
Need to use `startIndex + i` for `reverseMap` index.

```suggestion
		go func(startIndex int, payloads []*ledger.Payload) {
			minimap := make(map[flow.RegisterID]int, len(payloads))
			for i, payload := range payloads {
				key, err := payload.Key()
				if err != nil {
					minimaps <- result{nil, err}
					return
				}
				id, err := convert.LedgerKeyToRegisterID(key)
				if err != nil {
					minimaps <- result{nil, err}
					return
				}
				minimap[id] = startIndex + i
			}
			minimaps <- result{minimap, nil}
		}(start, payloadsCopy[start:end])
```